scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -323,127 +327,198 @@ public Pipeline getPipeline() {
-          } catch (Exception e) {
-            throw new RuntimeException(e);
+      // Create one iterator per shard
+      fileAssignments.forEach((resourceId, fileAssignment) -> {
+        final BucketMetadata<K, V> metadata = getMetadata().get(resourceId);
+
+        final int numBucketsInSource = metadata.getNumBuckets();
+        final int numShards = metadata.getNumShards();
+
+        // Since all BucketedInputs have a bucket count that's a power of two, we can infer
+        // which buckets should be merged together for the join.
+        for (int i = bucketId; i < numBucketsInSource; i += leastNumBuckets) {
+          for (int j = 0; j < numShards; j++) {
+            final ResourceId file = fileAssignment.forBucket(BucketShardId.of(i, j), metadata);
+            try {
+              iterators.add(fileOperations.iterator(file));
+            } catch (Exception e) {
+              throw new RuntimeException(e);
+            }
           }
         }
-      }
+      });
+
+      final Function<V, byte[]> keyBytesFn = getCanonicalMetadata()::getKeyBytes;
 
       // Merge-sort key-values from shards
       final UnmodifiableIterator<V> iterator =
           Iterators.mergeSorted(
               iterators,
               (o1, o2) ->
                   bytesComparator.compare(
-                      getMetadata().getKeyBytes(o1), getMetadata().getKeyBytes(o2)));
-      return new KeyGroupIterator<>(iterator, getMetadata()::getKeyBytes, bytesComparator);
+                      keyBytesFn.apply(o1), keyBytesFn.apply(o2)));
+      return new KeyGroupIterator<>(iterator, keyBytesFn, bytesComparator);
     }
 
     @Override
     public String toString() {
       return String.format(
-          "BucketedInput[tupleTag=%s, inputDirectory=%s, metadata=%s]",
-          tupleTag.getId(), inputDirectory, getMetadata());
+          "BucketedInput[tupleTag=%s, inputDirectories=[%s], metadata=%s]",
+          tupleTag.getId(),
+          inputDirectories.size() > 5 ?
+              inputDirectories.subList(0, 4) + "..." + inputDirectories.get(inputDirectories.size() - 1)
+              : inputDirectories,
+          getCanonicalMetadata());
     }
 
     private static class BucketedInputCoder<K, V> extends AtomicCoder<BucketedInput<K, V>> {
       private static SerializableCoder<TupleTag> tupleTagCoder =
           SerializableCoder.of(TupleTag.class);
-      private static ResourceIdCoder resourceIdCoder = ResourceIdCoder.of();
+      private static ListCoder<ResourceId> inputDirectoriesCoder =
+          ListCoder.of(ResourceIdCoder.of());
       private static StringUtf8Coder stringCoder = StringUtf8Coder.of();
       private static SerializableCoder<FileOperations> fileOpCoder =
           SerializableCoder.of(FileOperations.class);
-      private BucketMetadataCoder<K, V> metadataCoder = new BucketMetadataCoder<>();
+      private MapCoder<ResourceId, BucketMetadata<K, V>> metadataMapCoder = MapCoder.of(
```

Comment:
So I don't really like that I'm serializing the entire map of input dir -> metadata, but we do need to know how many buckets are in that particular directory when we iterate (i.e. if we're merging a directory with 2 buckets with a directory with 8 buckets) :/

Reply:
This is for `BucketedInput` in a `PCollection` right? It's not part of the job graph and should be OK.